libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/DefaultTracer.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.telemetry.tracing;
+
+import java.io.Closeable;
+import java.io.IOException;
+
+/**
+ *
+ * The default tracer implementation. This class implements the basic logic for span lifecycle and its state management.
+ * It also handles tracing context propagation between spans.
+ *
+ *
+ */
+public class DefaultTracer implements Tracer {
+
+    /**
+     * Key for storing current span
+     */
+    public static final String CURRENT_SPAN = "current_span";
+
+    static final String THREAD_NAME = "th_name";
+
+    private final TracingTelemetry tracingTelemetry;
+    private final TracerContextStorage<String, Span> tracerContextStorage;
+
+    /**
+     * Creates DefaultTracer instance
+     *
+     * @param tracingTelemetry tracing telemetry instance
+     * @param tracerContextStorage storage used for storing current span context
+     */
+    public DefaultTracer(TracingTelemetry tracingTelemetry, TracerContextStorage<String, Span> tracerContextStorage) {
+        this.tracingTelemetry = tracingTelemetry;
+        this.tracerContextStorage = tracerContextStorage;
+    }
+
+    @Override
+    public Scope startSpan(String spanName) {
+        Span span = createSpan(spanName, getCurrentSpan());
+        setCurrentSpanInContext(span);
+        addDefaultAttributes(span);
+        return new ScopeImpl(() -> endSpan());
```

Comment:
shouldn't it always pass the span created above to `endSpan()` method, instead of depending on `endSpan()` to get currentSpan and end.

Reply:
I misinterpreted's Rishabh's comment, this makes sense. Have addressed it.